### ISSUE TITLE
Looking forward to support for Swoole coroutines

### ISSUE THREAD

**Issue by clemask** [laravel/octane#765]
We really look forward to Octane supporting Swoole coroutines, which will bring higher performance and solve more business scenarios. We hope the official will include this in the roadmap as soon as possible.

**Comment by taylorotwell** [comment]
Feel free to work on it! 👍

**Comment by clemask** [comment]
Octane is a great project that is pushing Laravel in a new direction in the realm of the internet, making Laravel more powerful. However, there seem to be some issues with the way it's being handled in certain aspects. Octane's support for Swoole has excited many developers, including myself, envisioning a strong combination of Laravel's ecosystem with Swoole's coroutine advantages.

One crucial point that is not clearly stated in the official documentation is the inability to use Swoole's coroutine. This omission is significant because developers might unknowingly configure coroutine parameters, expecting Octane to work seamlessly with Swoole's coroutine features. However, Octane does not disable the creation of Swoole services with coroutine parameters, leading to subtle issues in the production environment. Problems like network connections being used by other coroutines and issues with Monolog log cycling detection can arise. In the end, developers are forced to disable coroutines as a quick solution, which is a severe incident.

I hope the official team promptly considers adding coroutine support to their roadmap. It's a fantastic feature that not only enhances Laravel but is also crucial for the broader PHP ecosystem.

**Comment by driesvints** [comment]
> I hope the official team promptly considers adding coroutine support to their roadmap.

This isn't a trivial thing to do. Right now we don't have the time to work on this ourselves but we'd love to see PR's kickstarting this.

**Comment by driesvints** [comment]
Welcoming prs 👍

**Comment by binaryfire** [comment]
@driesvints I've been active in the Hyperf community lately. I'll put the word out and see if any Swoole gurus can help.

**Comment by huangdijia** [comment]
If you want octane to support coroutine, you must first make database, redis, etc. support connection pooling.

**Comment by binaryfire** [comment]
I guess the problem is Laravel supports a lot of DB and cache backends and it’d be hard to implement pooling for them all…

@taylorotwell @driesvints @nunomaduro Would the core team be interested in having pooling for a select few drivers? Maybe just MySQL and Redis to start. It’d mean coroutines could only be enabled when using those drivers, but Hyperf can handle 100k req/second. Even if coroutines ‘only’ bumped Laravel’s concurrency to 10k-20k, that’d be nuts and create some serious buzz. That’s faster than Node…

FYI @huangdijia is a key Hyperf contributor: https://github.com/hyperf/hyperf/commits/master. He might be willing to help if you guys want to move ahead.

**Comment by foremtehan** [comment]
Can this be related? 👀

![image](https://github.com/laravel/octane/assets/53290883/e71f62b0-5f4d-40a9-bae9-19e7ad765270)

**Comment by binaryfire** [comment]
> Can this be related? 👀
> 
>

No that’s not related. Running a few tasks concurrently is different to workers that can properly handle multiple requests per second, which requires db connection pooling etc. I don’t think there’s much interest from the core team in increasing Laravel’s performance beyond the current Octane implementation, which only helps by keeping the app booted in memory. It doesn’t use Swoole coroutines at all.

If you want to handle 40k - 100k req/s in real life workloads instead of Octane’s 500-ish, you’ll need to either run a bunch of Octane instances behind a load balancer, or look at a proper coroutine framework like Hyperf. The performance difference is night and day. 

People who say ~500 req/s is enough just aren’t building certain kinds of apps. Eg. a log ingestion app, API gateways etc. Even a core team member has written articles about adding Go to a Laravel app stack to handle endpoints which receive high numbers of requests, which clearly demonstrates Laravel’s performance limitations.

Shipping more features seems to be a higher priority than a major upgrade to performance. This kind of change, and the refactoring that would be required (eg. adding db connection pooling) won’t ever come from a third party PR. It would need be be driven by the core team.

Hyperf is heavily inspired by Laravel so there are plenty of code references there the core team could look at.

**Comment by huangdijia** [comment]
> > Can this be related? 👀
> 
> > 
> 
> >
> 
> 
> 
> No that’s not related. Running a few tasks concurrently is different to workers that can properly handle multiple requests per second, db connection pooling etc. I don’t think there’s much interest from the core team in increasing Laravel’s performance beyond Octane, which only helps by keeping the app booted in memory. 
> 
> 
> 
> So if you want to handle 40k - 100k req/s in real life workloads instead of Octane’s 500-ish, you’ll need to either run a bunch of Octane instances behind a load balancer, or look at a PHP framework like Hyperf.
> 
> 
> 
> People who say 500 req/s is enough just aren’t building certain kinds of apps. Eg. a log ingestion app, an API gateway etc. Even a member of the core team has written articles about adding Go to a Laravel app stack to handle endpoints which receive high numbers of requests, which is saying something.
> 
> 
> 
> Shipping more features seems to be a higher priority than any major upgrades to performance. This kind of change and the refactoring (eg. adding db connection pooling) won’t come from a third party PR. It would need be be driven by the core team.

The design idea of octane is still based on fpm,To get better performance, you need to redesign

**Comment by SMFloris** [comment]
Managed to get Laravel working with multiple Swoole coroutines in one process today. I still have allot of cleaning up to do before a PR, but I am confident it could be ok. Currently we got it working for a custom Kernel that processes sqs jobs for aws lambda, but to get it working for the http kernel is trivial. By Friday, I'll have it up and running in multiple processes as well (probably already works, but needs further testing).

@taylorotwell  , @nunomaduro , is this still of interest?

**Comment by fannyfan414** [comment]
> Managed to get Laravel working with multiple Swoole coroutines in one process today. I still have allot of cleaning up to do before a PR, but I am confident it could be ok. Currently we got it working for a custom Kernel that processes sqs jobs for aws lambda, but to get it working for the http kernel is trivial. By Friday, I'll have it up and running in multiple processes as well (probably already works, but needs further testing).
> 
> @taylorotwell , @nunomaduro , is this still of interest?

This will be the most long-awaited and amazing feature for all highload Laravel projects!

**Comment by SMFloris** [comment]
> This kind of change, and the refactoring that would be required (eg. adding db connection pooling) won’t ever come from a third party PR. It would need be be driven by the core team.

Interestingly enough I did run into this today with testing multiple processes with multiple coroutines each. Database pooling is kinda of a solved problem upstream in openswoole at least (I am using openswoole atm).

https://github.com/openswoole/openswoole/blob/master/example/src/Coroutine/MySQLClientPool.php

Also available: Redis, Postgres.

Going to implement a mysql and redis pool next, which seems to be the last step I need to get the whole thing running blazingly fast. It is actually quite amazing how fast everything is.

**Comment by binaryfire** [comment]
@SMFloris Nice! I’d suggest using Swoole rather than Openswoole for testing - all Swoole development happens there. The whole “security” thing was blown completely out of proportion by the Openswoole guy who now just repackages Swoole by himself to try and steal some brand recognition. 

Also one of the core Swoole maintainers (@deminy) has said he’s happy to answer questions to help: https://github.com/swoole/swoole-src/pull/4330#issuecomment-1837271923

Re: the performance difference - yeah I’ve used Hyperf for several high load projects and the difference is night and day. It can handle approx 40x more req/s than Laravel Octane thanks to being built with coroutines from the ground up. Pretty comparable to Go. Companies are building game server backends with it, which is pretty nuts for a PHP framework.

IMHO properly implementing coroutines in Laravel would open up the framework to a whole group of devs who would never have considered PHP otherwise.

@huangdijia is a core Hyperf contributor and might be willing to review your PR too

I’d also suggest posting some benchmarks with your PR to show the performance benefits.

**Comment by SMFloris** [comment]
@binaryfire thanks for your input! Will use Swoole going forward.

I was just thinking that the level of speed is at the level of Go as well.

What I'll do after implementing the pooling connections as well (I'm trying not to hack to much, doing things the Laravel way as much as possible) is that I'll create a repo with a docker compose so that anyone can run the benchmark easily.

What sort of things would be nice to benchmark? I'm thinking of a route that simply returns an OK status and a route that does some inserts or reads from the database. I personally want to compare: simple Laravel behind nginx, laravel octane as it currently is and laravel octane with all the bells and whistles from coroutines.

**Comment by SMFloris** [comment]
This weekend I made some headway;

Turns out there are allot more stuff to wire together than I thought initially.

- First of all, Swoole has some proxy classes for connection pools (i.e. PDOProxy, PDOStatementProxy) and the Connection classes in Laravel need to have their methods' definitions rewritten to also include these. Largely, though, I was able to use the existing MysqlConnection class. I don't know if there already is an interface for the PDO class.
- Then, there are allot of abstractions to take care of in octane specifically; not allot, but enough to give me a hard time.
- Some things will require some more work (like, Octane::concurrently and the task worker)

**Comment by evan361425** [comment]
Why not using [coroutine hooks](https://openswoole.com/docs/modules/swoole-runtime-flags)?

It don't need to change code base, just hooking the PHP C code interface.
We can enable it by calling `\OpenSwoole\Runtime::enableCoroutine();` before worker start.

I've tried it as issue #906 mentioned but encountered some issues in binding process.

**Comment by xwiz** [comment]
@SMFloris can you possibly push your changes so another person can pick up from where you are at or assist.

**Comment by clemask** [comment]
> This weekend I made some headway;
> 
> Turns out there are allot more stuff to wire together than I thought initially.
> 
> * First of all, Swoole has some proxy classes for connection pools (i.e. PDOProxy, PDOStatementProxy) and the Connection classes in Laravel need to have their methods' definitions rewritten to also include these. Largely, though, I was able to use the existing MysqlConnection class. I don't know if there already is an interface for the PDO class.
> * Then, there are allot of abstractions to take care of in octane specifically; not allot, but enough to give me a hard time.
> * Some things will require some more work (like, Octane::concurrently and the task worker)
Is there any new progress?

**Comment by clemask** [comment]
@SMFloris Is there any new progress?

**Comment by osbre** [comment]
To anyone being passionate about this feature, I recommend reading https://openswoole.com/article/isolating-variables-with-coroutine-context.

As I see it, simply flipping the switch to do `Runtime::enableCoroutine` and implementing connection pooling would result in a buggy behaviour throughout the app, due to the number of static variables that would essentially get shared between all of the coroutines. You can look at [how Octane currently has to reset the state of the container inside each worker before processing a request.](https://github.com/laravel/octane/blob/2.x/src/Concerns/ProvidesDefaultConfigurationOptions.php)

There appears to be an architectural problem with static variables which wouldn't let us use coroutines as of right now. All of that state which is being reset has to be stored in the coroutine context rather than globally.

I'll also look into this later.

Upd:
Not just static variables, many of the container instances have to be retrieved from the coroutine state rather than just reset.

**Comment by huangdijia** [comment]
https://laravel-hyperf.com/ @binaryfire

**Comment by clemask** [comment]
> https://laravel-hyperf.com/ @binaryfire

Can it be migrated safely, easily and quickly?

**Comment by harryqt** [comment]
No yet. The project is at alpha stage.

**Comment by huangdijia** [comment]
> > https://laravel-hyperf.com/ @binaryfire
>
> Can it be migrated safely, easily and quickly?

Currently the laravel ecosystem is not supported, such as horizon, telescope, etc.

**Comment by binaryfire** [comment]
> No yet. The project is at alpha stage.

Hyperf is a completely different framework. The starter project Deeka linked to is just arranged in a way that’ll be familiar to Laravel devs. Directory structure etc.

You can’t swap out Laravel for Hyperf in an exisiting project. It would require significant rewriting.

**Comment by Cluster2a** [comment]
> > No yet. The project is at alpha stage.
> 
> Hyperf is a completely different framework. The starter project Deeka linked to is just arranged in a way that’ll be familiar to Laravel devs. Directory structure etc.
> 
> You can’t swap out Laravel for Hyperf in an exisiting project. It would require significant rewriting.

We migrated a module from our Laravel-Monolith to Hyperf, to run it as a separate Service.
We only had to setup the bootstrapping (Services/middlewares and so on).
For the business logic, we copied over the models, the controllers and everything was working almost out of the box.

**Comment by binaryfire** [comment]
> We migrated a module from our Laravel-Monolith to Hyperf, to run it as a separate Service.
> We only had to setup the bootstrapping (Services/middlewares and so on).
> For the business logic, we copied over the models, the controllers and everything was working almost out of the box.

Yeah it's great for microservices. We have several high traffic services using Hyperf - the performance difference is massive. The framework was heavily inspired by Laravel and uses a forked coroutine-friendly version of Eloquent so it's easy for Laravel devs to pick up. Things like models are easy to move over but there are many additional considerations when building complex apps eg. the way Context works.

Not everything in the ecosystem can be ported over eg Livewire: https://github.com/hyperf/hyperf/issues/5517. We've gone all in on Livewire so unfortunately that makes Hyperf a no-go for our core apps. Also there's no first-party auth package since the framework is geared towards microservices.  But it's our first choice for anything we have to break out.

**Comment by clemask** [comment]
> > > https://laravel-hyperf.com/ @binaryfire
> > 
> > 
> > Can it be migrated safely, easily and quickly?
> 
> Currently the laravel ecosystem is not supported, such as horizon, telescope, etc.

I don’t know the usage of Horizon and Telescope. Compared with those using Laravel, it should be relatively low. It would be very exciting if it could support the migration of Laravel projects. @huangdijia

**Comment by huangdijia** [comment]
> I don’t know the usage of Horizon and Telescope. Compared with those using Laravel, it should be relatively low. It would be very exciting if it could support the migration of Laravel projects. @huangdijia 

Welcome to learn about and use Hyperf. It can be accessed via the documentation (https://hyperf.wiki) to Learn about it.

**Comment by albertcht** [comment]
Hello everyone,

I'm the creator of the [Hypervel](https://hypervel.org) project. This project is currently in its final stage before the first dev release. I've migrated most components from Laravel to make them coroutine-friendly while striving to maintain a similar development experience to Laravel.

From my perspective, implementing full coroutine support in the Laravel framework presents significant challenges for several reasons:

1. Laravel's components weren't designed with coroutines in mind. This means all states would need to be isolated within the coroutine context, as @osbre mentioned. Laravel Octane's approach of resetting states after each request doesn't address this fully. For instance, if a singleton object needs to be shared across coroutines, the reset mechanism becomes ineffective.

2. Swoole introduced a `max_concurrency` setting ([see here](https://github.com/swoole/swoole-src/pull/4330)) to enable Laravel Octane to support coroutines with limited restrictions. When `max_concurrency` is set to 1, the worker won't accept new requests until the current one is completed. While this ensures coroutine safety and allows for state resets, it essentially makes coroutines an alternative concurrency feature in Laravel without improving QPS performance, as a worker can still only handle one request at a time.

3. Even if Laravel Octane were to support full coroutines in the future, it would necessitate significant breaking changes in Laravel components. All states would need to be safely stored using context, which would have a substantial impact on the third-party ecosystem.

The complexity of migrating from Laravel to Hypervel depends on the extent of state management and third-party packages used in your project. If you're considering migration, keep in mind:

1. Hypervel is a distinct framework, so a cost-free migration process is unrealistic.
2. Laravel packages can't be used directly in Hypervel due to coroutine safety issues.
3. The Hypervel ecosystem is still in its early stages. If you require specific Laravel packages that haven't been migrated to Hypervel yet, you may need to do it yourself.

In conclusion, I highly recommend trying Hypervel for new projects with intensive I/O requirements. However, if you're planning to migrate an existing Laravel project, be prepared for necessary modifications and consider the ecosystem implications carefully.

> For current Hyperf users, you can also refer to [firendsofhyperf](https://github.com/friendsofhyperf) for some useful packages.

**Comment by foremtehan** [comment]
> Even if Laravel Octane were to support full coroutines in the future, it would necessitate significant breaking changes in Laravel components. All states would need to be safely stored using context, which would have a substantial impact on the third-party ecosystem.

No need to introduce breaking changes, They need to fork laravel/framework and continue developing the fork version. @themsaid

**Comment by fakharksa** [comment]
Trying to integrate Swoole into PHP-FPM-based (PHP-) Frameworks (like, Laravel) is entirely (and, philosophically) a wrong direction to go.

People who want to enjoy modern (CSP based) Asynchronous PHP-Programming will need to shift the paradigm of their thought Process from PHP-FPM based frameworks to "Asynchronous Programming Paradigm". 

They (PHP-FPM and Swoole) both utilise entirely different OS-concepts and entirely different programming concepts, and entirely different Application Architectures. Swoole is just like Go, VertX, Netty, Erlang, Tornado etc.

**Comment by foremtehan** [comment]
Swoole introduced: [Swoole\Thread](https://github.com/swoole/swoole-src/releases/tag/v6.0.0)

Could it help?

**Comment by proilyxa** [comment]
https://github.com/proilyxa/lara-thread
I posted an example here of how you can use swoole threads with laravel

**Comment by mikkpokk** [comment]
> Swoole introduced: [Swoole\Thread](https://github.com/swoole/swoole-src/releases/tag/v6.0.0)
> 
> Could it help?

I don’t see how multi-thread support can help with coroutine support.

Coroutines require fundamental changes at the framework level. The current implementation relies on process blocking.

For example, when you fetch a user from the User model, the request (worker) is blocked until the user is retrieved from the database.

Of course, you can fetch the user and process it in a task worker, but if your HTTP response directly depends on the retrieved user data, you cannot delegate that to a task worker - the job has to be done at the worker level.

To simplify and make the problem more relatable, you can think of a task worker as a classical Laravel queue task and a worker as a classical Laravel HTTP request. Coroutines, on the other hand, are something in between a queue and an HTTP request - they can communicate with each other when necessary.

Multi-threading task workers are simply a new ability to load-balance specific heavy tasks across multiple threads. However, every worker and task worker remains single-threaded otherwise.

**Comment by superbiche** [comment]
@SMFloris hey, do you think you could publish your WIP in a repo so people can help work on it?

**Comment by mikkpokk** [comment]
> [@SMFloris](https://github.com/SMFloris) hey, do you think you could publish your WIP in a repo so people can help work on it?

It is not Octane related. It's Laravel related. There's obviously no intention to support Swoole with it's all powers on Laravel. It requires framework rewrite. Migrate your project to [Hypervel](https://hypervel.org/) instead.

**Comment by eblanshey** [comment]
I'm new to Swoole and async PHP, so forgive my ignorance, but the [Swoole docs](https://wiki.swoole.com/en/#/runtime?id=function-prototype) indicate automatically enabling non-blocking operations via `Swoole\Runtime::enableCoroutine(SWOOLE_HOOK_ALL)`. It includes PDO, Redis, and much more. Is there a reason this cannot be utilized out of the box in Laravel Octane?

**Comment by albertcht** [comment]
> I'm new to Swoole and async PHP, so forgive my ignorance, but the [Swoole docs](https://wiki.swoole.com/en/#/runtime?id=function-prototype) indicate automatically enabling non-blocking operations via `Swoole\Runtime::enableCoroutine(SWOOLE_HOOK_ALL)`. It includes PDO, Redis, and much more. Is there a reason this cannot be utilized out of the box in Laravel Octane?

You can refer to this comment: https://github.com/laravel/octane/issues/765#issuecomment-2408769353

It's basically because Laravel doesn't handle states at the coroutine level. This can cause states bleeding when switching between coroutines.

**Comment by proilyxa** [comment]
Hey! Great news! Coroutines are here in Laravel!

https://github.com/yangusik/laravel-spawn

**Comment by Cluster2a** [comment]
You might want to check on https://true-async.github.io/en/
Should be stable by November 2026

**Comment by fakharanwarcom-prog** [comment]
> I'm new to Swoole and async PHP, so forgive my ignorance, but the [Swoole docs](https://wiki.swoole.com/en/#/runtime?id=function-prototype) indicate automatically enabling non-blocking operations via `Swoole\Runtime::enableCoroutine(SWOOLE_HOOK_ALL)`. It includes PDO, Redis, and much more. Is there a reason this cannot be utilized out of the box in Laravel Octane?

Answer from @albertcht and the recommendation from @Cluster2a to use TrueAsynch over Swoole is not correct.

In modern "In-memory" Concurrent Eventing Architectures, Application Servers (Engines like Swoole) are Stateful with purpose. So retaining / maintaining some global state across requests is not "Leaking (a bug / issue)" of State (data), instead it is "Sharing" of some global state inside memory.

It is this statefulness the whole Framework does not have to reboot on every User Request, as it happens in case of Laravel which runs on PHP-FPM.

PHP-FPM is stateless which means PHP-FPM is based on "shared-nothing" Architectural concept. PHP-FPM's worker does not serve more than one requests concurrently, instead it serves one request after another. Contrary to this, Swoole Engine multiplex Requests per Worker which is Laravel Octane (even when running on Swoole) does not want to support so that its PHP Packages remain compatible with "shared nothing" architecture (each request bootstrap everything and destroys everything before another request occupy the Swoole Worker, and Swoole Team specailly made this "mode of op." inside its engine for Laravel Octane (and other PHP-FPM base traditional PHP-Frameworks).

This mode decreses the level of Swoole's original concurrency (multiple-Requests to run concurreently within each Worker).

TrueAsynch uses libuv (just like ReactPHP and AMPHP and Revolt and Node.js do) which does not support what Swoole supports; Multiple Event-loops; one Event-loop per Worker along with Pre-emptive Scheduler and support to fastest and latest Linux algo io_uring for Eventing. (This makes Swoole 6.2.2 blazing fast; 3+ times faster than Node.js and Golang).

Therefore, i strongly recommend to learn and use Swoole directly without relying on Third-parties in the middle of Swoole and you. No wrapper is better than Swoole itself. (Plus, you get single-point of Support, instead of reliance on multiple different Development Teams). Swoole has superset of features like Locks, Swoole Table (Hash Table), Worker Pools, channels, multi-Threading, multiple OS Processes / Pools etc.

**Comment by EdmondDantes** [comment]
> Multiple Event-loops; one Event-loop per Worker along with Pre-emptive Scheduler and support to fastest and latest Linux algo io_uring for Eventing. (This makes Swoole 6.2.2 blazing fast; 3+ times faster than Node.js and Golang

In synthetic benchmarks, the TrueAsync server outperformed Swoole. However, I have never seen Swoole being three times faster. In fact, that shouldn't even be theoretically possible. io_uring alone does not provide that kind of performance gain, especially in typical HTTP workloads.
And in real-world applications, the performance difference between these technologies is even smaller. So this is purely marketing.

> “Swoole has a pre-emptive scheduler.”

An important caveat is needed here. Swoole coroutines are fundamentally cooperative: context switching naturally occurs at yield/I/O points. The preemptive scheduling option uses periodic PHP VM interrupts to prevent a single CPU-bound coroutine from monopolizing a worker indefinitely. This is not equivalent to an OS preemptive thread scheduler, nor does it turn PHP coroutines into parallel computations. Simply calling it a “pre-emptive scheduler” without explaining this limitation is misleading.

> Multiple Event-loops;

Multiple event loops in a single thread? That makes no sense. Multiple event loops per process? That's a standard architecture.
